BackEnd/source/server.js

```diff
@@ -0,0 +1,257 @@
+const express = require('express');
+
+const app = express();
+
+const state = {
+  gameBoardValues: [],
+  currentPlayer: 'red',
+  gameBoardColumns: 0,
+  gameBoardRows: 0,
+};
+
+function setupGameBoardValuesArray(row, column) { // PURE FUNCTION
+  const customGameBoardValues = [];
+  for (let i = 0; i < column; i += 1) {
+    const tempArray = [];
+    for (let j = 0; j < row; j += 1) {
+      tempArray.push(null);
+    }
+    customGameBoardValues.push(tempArray);
+  }
+  return customGameBoardValues;
+}
+
+function resetBoardArray(row, column, gameBoardArray) { // PURE FUNCTION
+  const customGameBoardValues = gameBoardArray.slice;
+  for (let i = 0; i < row; i += 1) {
+    for (let j = 0; j < column; j += 1) {
+      customGameBoardValues[i][j] = null;
+    }
+  }
+  return customGameBoardValues;
+}
+
+function updateStateValues(newStateValues) {
+  state.gameBoardValues = newStateValues.gameBoardValues;
+  state.currentPlayer = newStateValues.currentPlayer;
+  state.gameBoardColumns = newStateValues.gameBoardColumns;
+  state.gameBoardRows = newStateValues.gameBoardRows;
+}
+
+function changePlayer(currentPlayerColour) { // PURE FUNCTION
+  if (currentPlayerColour === 'yellow') {
+    return 'red';
+  }
+  return 'yellow';
+}
+
+function placeCounter(columnIndex) {
+  for (let i = state.gameBoardValues[columnIndex].length - 1; i >= 0; i -= 1) {
+    if (state.gameBoardValues[columnIndex][i] == null) {
+      state.gameBoardValues[columnIndex][i] = state.currentPlayer;
+
+      state.currentPlayer = changePlayer(state.currentPlayer);
+
+      const counterPos = `${columnIndex}-${i}`;
+      return counterPos;
+    }
+  }
+  return null;
+}
+
+function checkColumnWin(columnIndex, gameBoardArray) { // PURE FUNCTION
+  let fourInARowCount = 0;
+  for (let i = gameBoardArray[columnIndex].length - 1; i >= 0; i -= 1) { // Loops through each row
+    if ((i + 1) >= 4 && gameBoardArray[columnIndex][i] !== null) {
```

Comment:
It looks like the `(i+1)>=4` condition probably belongs as the conditional part of the for loop definition
```js
for (let i = gameBoardArray[0].length - 1; i >= 3; i -= 1)
```
Beyond that, rather than wrapping the whole thing in an if statement, it's preferred to break early like this:
```js
if (gameBoardArray[columnIndex][i] === null) {
    break
}
// rest of code follows here
```
Deeply nested code is harder to follow, because you have to keep at the back of the mind the context you're going to go back to when the loop or conditional ends. So if you can get an if-statement over and done with up-front, that makes it easier to understand.